README.md
[![Yii2](https://img.shields.io/badge/required-Yii2_v2.0.40-blue.svg)](https://packagist.org/packages/yiisoft/yii2)
[![Downloads](https://img.shields.io/packagist/dt/wdmg/yii2-comments.svg)](https://packagist.org/packages/wdmg/yii2-comments)
[![Packagist Version](https://img.shields.io/packagist/v/wdmg/yii2-comments.svg)](https://packagist.org/packages/wdmg/yii2-comments)
![Progress](https://img.shields.io/badge/progress-ready_to_use-green.svg)
[![GitHub license](https://img.shields.io/github/license/wdmg/yii2-comments.svg)](https://github.com/wdmg/yii2-comments/blob/master/LICENSE)

<img src="./docs/images/yii2-comments.png" width="100%" alt="Yii2 Comments Module" />

# Yii2 Comments Module
Tree comments system for Yii2 with widget. The functionality of the module allows you to manage user comments with pre-moderation or automatic approval for registered users.

This module is an integral part of the [Butterfly.СMS](https://butterflycms.com/) content management system, but can also be used as an standalone extension.

Copyrights (c) 2019-2023 [W.D.M.Group, Ukraine](https://wdmg.com.ua/)

# Requirements 
* PHP 5.6 or higher
* Yii2 v.2.0.40 and newest
* [Yii2 Base](https://github.com/wdmg/yii2-base) module (required)
* [Yii2 SelectInput](https://github.com/wdmg/yii2-selectinput) widget
* [Yii2 Users](https://github.com/wdmg/yii2-users) module (optionaly)

# Installation
To install the module, run the following command in the console:

`$ composer require "wdmg/yii2-comments"`

After configure db connection, run the following command in the console:

`$ php yii comments/init`

And select the operation you want to perform:
  1) Apply all module migrations
  2) Revert all module migrations

# Migrations
In any case, you can execute the migration and create the initial data, run the following command in the console:

`$ php yii migrate --migrationPath=@vendor/wdmg/yii2-comments/migrations`

# Configure
To add a module to the project, add the following data in your configuration file:

    'modules' => [
        ...
        'comments' => [
            'class' => 'wdmg\comments\Module',
            'routePrefix' => "admin",
            'defaultController' => "admin/comments/default",
            'baseRoute' => "/comments",
            'defaultListView' => "@vendor/wdmg/yii2-comments/widgets/views/_list",
            'defaultFormView' => "@vendor/wdmg/yii2-comments/widgets/views/_form",
            'editCommentTimeout' => 300,
            'deleteCommentTimeout' => 3600,
            'newCommentsModeration' => true,
            'approveFromRegistered' => true
        ],
        ...
    ],

# Usecase
Use the CommentsWidget::widget() to render comment`s list by context and target.

    <?php
        echo \wdmg\comments\widgets\CommentsWidget::widget([
            'context' => 'page-comments',
            'target' => 'site/index'
        ]);
    ?>

# Routing
Use the `Module::dashboardNavItems()` method of the module to generate a navigation items list for NavBar, like this:

    <?php
        echo Nav::widget([
        'options' => ['class' => 'navbar-nav navbar-right'],
            'label' => 'Modules',
            'items' => [
                Yii::$app->getModule('comments')->dashboardNavItems(),
                ...
            ]
        ]);
    ?>

# Status and version [ready to use]
* v.2.0.0 - Update copyrights, fix nav menu
* v.1.0.3 - Added counter stats method
* v.1.0.2 - Fixed mysql syntax error: 1055 for >= MySQL 5.7
* v.1.0.1 - Update README.md


CHANGELOG.md
Changelog
=========

## 2.0.0 (2023-06-25)
 * Update copyrights, fix nav menu

## 1.0.3 (2021-05-10)
 * Added counter stats method
 
## 1.0.2 (2021-01-17)
 * Fixed mysql syntax error: 1055 for >= MySQL 5.7
 
## 1.0.1 (2020-06-14)
 * Added assets (gitignore fix)
 * Update README.md

## 1.0.0 (2020-06-07)
 * Default controller for frontend
 * View comment, edit comment form
 * Form template in CommentsWidget rebuild.
 * Added _form and _list views for widget
 * Added assets and CommentsWidget
 * Migrations rebase, refactoring model
 
## 0.0.11 (2020-03-25)
 * Added pagination
 * Up to date dependencies
 
## 0.0.10 (2019-12-10)
 * Fixed deprecated class declaration

## 0.0.9 (2019-07-10)
 * Added extra options to composer.json and navbar menu icon

## 0.0.8 (2019-06-08)
 * Added choice param for non interactive mode

## 0.0.7 (2019-06-07)
 * Module refactoring
 
## 0.0.6 (2019-06-06)
 * Module transferred to base module interface. Update Yii2 version.
 
## 0.0.5 (2019-06-01)
 * Bugfix, routing and refactoring
 
## 0.0.4 (2019-05-30)
 * Update Yii2 version
 
## 0.0.3 (2019-05-28)
 * Added module name and description, translations. Routing fix.
 
## 0.0.2 (2019-03-25)
 * Added base models and CRUD
 * Added base translations
 * Added composer package
 * Added console controller
 
## 0.0.1 (2019-03-24)
 * Added bootstrap, base module
 * Added base migrations

Module.php
<?php

namespace wdmg\comments;

/**
 * Yii2 Comments
 *
 * @category        Module
 * @version         2.0.0
 * @author          Alexsander Vyshnyvetskyy <alex.vyshnyvetskyy@gmail.com>
 * @link            https://github.com/wdmg/yii2-comments
 * @copyright       Copyright (c) 2019 - 2023 W.D.M.Group, Ukraine
 * @license         https://opensource.org/licenses/MIT Massachusetts Institute of Technology (MIT) License
 *
 */

use wdmg\helpers\ArrayHelper;
use Yii;
use wdmg\base\BaseModule;

/**
 * comments module definition class
 */
class Module extends BaseModule
{
    /**
     * {@inheritdoc}
     */
    public $controllerNamespace = 'wdmg\comments\controllers';

    /**
     * {@inheritdoc}
     */
    public $defaultRoute = 'comments/index';

    /**
     * @var string, the name of module
     */
    public $name = "Comments";

    /**
     * @var string, the description of module
     */
    public $description = "Tree comments system";

    /**
     * The default frontend controller
     *
     * @var string
     */
    public $defaultController = "admin/comments/default";

    /**
     * The default routes to default controller in frontend.
     *
     * @var string or array, use "/" - for root
     */
    public $baseRoute = "/comments";

    /**
     * Default layout for listing comments in frontend.
     *
     * @var string
     */
    public $defaultListView = '@vendor/wdmg/yii2-comments/widgets/views/_list';

    /**
     * Default layout for outputting the form for adding / editing comments to the frontend.
     *
     * @var string
     */
    public $defaultFormView = '@vendor/wdmg/yii2-comments/widgets/views/_form';

    /**
     * The time during which the comment is available for editing by its author.
     *
     * @var int, in seconds. Use `0` for disable option.
     */
    public $editCommentTimeout = 300; // (5 min.)

    /**
     * The time during which the comment is available for removal by the author.
     *
     * @var int, in seconds. Use `0` for disable option.
     */
    public $deleteCommentTimeout = 3600; // (1 hour)

    /**
     * Moderate all new comments.
     *
     * @var bool, if `true` - all new comments be marked as awaiting moderation status.
     */
    public $newCommentsModeration = true;

    /**
     * Auto approve all new comments from registered users.
     *
     * @var bool, if `true` - all new comments from registered users be marked as published
     */
    public $approveFromRegistered = true;

    /**
     * @var string the module version
     */
    private $version = "2.0.0";

    /**
     * @var integer, priority of initialization
     */
    private $priority = 8;

    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();

        // Set version of current module
        $this->setVersion($this->version);

        // Set priority of current module
        $this->setPriority($this->priority);

    }

    /**
     * {@inheritdoc}
     */
    public function dashboardNavItems($options = null)
    {
        $items = [
            'label' => $this->name,
            'url' => [$this->routePrefix . '/'. $this->id],
            'icon' => 'fa fa-fw fa-comments',
            'active' => in_array(\Yii::$app->controller->module->id, [$this->id])
        ];

	    if (!is_null($options)) {

		    if (isset($options['count'])) {
			    $items['label'] .= '<span class="badge badge-default float-right">' . $options['count'] . '</span>';
			    unset($options['count']);
		    }

		    if (is_array($options))
			    $items = ArrayHelper::merge($items, $options);

	    }

	    return $items;
    }

    /**
     * {@inheritdoc}
     */
    public function bootstrap($app)
    {
        parent::bootstrap($app);

        // Configure comments component
        $app->setComponents([
            'comments' => [
                'class' => \wdmg\comments\components\Comments::class
            ]
        ]);

        // UrlManager rules for frontend
        if (!$this->isBackend()) {
            $app->getUrlManager()->addRules([
                [
                    'pattern' => (($this->baseRoute) ?  $this->baseRoute : '/comments') . '/<action>',
                    'route' => (($this->defaultController) ?  $this->defaultController : 'comments/default') . '/<action>'
                ]
            ], true);
        }

    }
}